### PR TITLE
Boundary conditions

### PR DESCRIPTION
**Pull request by arthbous**
Tested the boundary conditions XBoundaries, YBoundaries, ZBoundaries.

The rest should work fine but we will have to test it directly on the PNP, because they are very specific.

Also added test_bc, in CMake and problems.

### CODE REVIEW

File: problems/test_bc/test_bc.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+int main()
+{
+
+  parameters["linear_algebra_backend"] = "Eigen"; // or uBLAS
+  parameters["allow_extrapolation"] = true;
+
+  // Create mesh and function space
+  dolfin::Point p0( -10.0, -10.0, -10.0);
+  dolfin::Point p1(  10.0,  10.0,  10.0);
+  dolfin::BoxMesh mesh(p0, p1, 10, 10, 5);
```
Comment (maximilianmetti) [inline problems/test_bc/test_bc.cpp]: You didn't use my ``domain_build()`` function!  I feel useless!!

File: problems/test_bc/test_bc.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  Poisson::FunctionSpace V(mesh);
+
+  // Define boundary condition
+  Constant ux(0.0);
+  XBoundaries bdary_x(10.0);
```
Comment (maximilianmetti) [inline problems/test_bc/test_bc.cpp]: What do you think if we simplify this to read in a list?
Something like:
```
unsigned int coordinate_index_array[4] = {0,0,2,2};
double coordinate_value_array[4] = {-L_x,L_x,L_z,L_y};
int direction_array[4] = {-1,1,-1,1};
Boundaries bndry( 
  &coordinate_index_array,
  &coordinate_value_array,
  &direction_array
);
```
I think the arrays are self-explanatory except ``direction_array``, which determines whether we are setting a lower bound (``-1``) or an upper bound (``+1``).

What do you think?  The function would return an array of ``DirichletBS``s.

**Comment by arthbous** [comment]
Sounds good to me!

**Comment by maximilianmetti** [comment]
did you make these changes or are you waiting for me to do it?

**Comment by arthbous** [comment]
Sorry I did not make these changes, I thought you wanted to do it. 

I can do it, but I don't understand how putting these data in a vector would help...

**Comment by maximilianmetti** [comment]
I was thinking that it combines more functionality into the method.  That way you can create all Dirichlet BCs at once and store them in a single object

File: problems/test_bc/test_bc.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  DirichletBC bc2(V, uy, bdary_y);
+
+  Constant uz(4.0);
+  ZBoundaries bdary_z(10.0);
+  DirichletBC bc3(V, uz, bdary_z);
+
+  std::vector<const DirichletBC*> bcs;
+  bcs.push_back(&bc1);
+  bcs.push_back(&bc2);
+  bcs.push_back(&bc3);
```
Comment (maximilianmetti) [inline problems/test_bc/test_bc.cpp]: This ``bcs`` object is what I propose is returned, so that lines 57-71 are inside the function call.

**Comment by arthbous** [comment]
I think you actually make a lot of sens, I will write that today

File: problems/test_bc/test_bc.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  dolfin::BoxMesh mesh(p0, p1, 10, 10, 5);
+
+  Poisson::FunctionSpace V(mesh);
+
+
+  // ##############################################################################
+  // ### First test to see if the boundary conditions works
+  // ##############################################################################
+  // dolfin::MeshFunction boundary_parts;
+  dolfin::MeshFunction<size_t> boundary_parts(mesh, mesh.topology().dim()-1);
```
Comment (maximilianmetti) [inline problems/test_bc/test_bc.cpp]: May I suggest: ``dolfin::FacetFunction<std::size_t> boundary(mesh);``

File: include/boundary_conditions.h [hunk 5]
```diff
@@ -11,104 +11,73 @@
...
+  bool inside(const dolfin::Array<double>& x, bool on_boundary) const;
 };
 
+class dielectricChannel : public dolfin::SubDomain
+{
+  double Lz;
+public:
+    dielectricChannel(double _Lz);
+    bool inside(const dolfin::Array<double>& x, bool on_boundary) const;
+};
```
Comment (maximilianmetti) [inline include/boundary_conditions.h]: We can remove these for now.  They're pretty problem specific.

File: problems/test_bc/test_bc.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  boundary_parts.set_all(0);
+
+
+  XBoundaries bdary_x(-10.0);
```
Comment (maximilianmetti) [inline problems/test_bc/test_bc.cpp]: Is this needed to set the ``public`` values for the ``XBoundaries``?  I'm wondering if we can get around this somehow...

File: include/boundary_conditions.h [hunk 7]
```diff
@@ -11,104 +11,73 @@
...
-// Before it was: class DirichletBoundary : public SubDomain
-class EastWestDomain: public SubDomain
-{
-    bool inside(const Array<double>& x, bool on_boundary) const
-    {
-        // Before it was: return on_boundary && (x[0] < -Lx+2.*DOLFIN_EPS or x[0] > Lx-2.*DOLFIN_EPS);
-        // Arthur changed it to:
-        return on_boundary && (x[0] < -Lx+DOLFIN_EPS || x[0] > Lx-DOLFIN_EPS); //(is that ok ?)
-    }
+class XBoundaries: public dolfin::SubDomain {
```
Comment (maximilianmetti) [inline include/boundary_conditions.h]: I think we can simplify this by being more careful about how we setup the class.  I'll take a look at an implementation following [this post](http://stackoverflow.com/questions/120876/c-superclass-constructor-calling-rules)

**Comment by arthbous** [comment]
I don't see any EastWestDomain
